---
title: "Differential expression (DE) Per Cell Type for sc-RNA-Seq Analysis in 10X Genomics data"
author: "Antonia Chroni for SJCRH DNB_BINF_Core"
papersize: a4
fontsize: 11pt
links-as-notes: true
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
    df_print: paged
    code_folding: hide
    toc_depth: 2
    highlight: tango
    number_sections: TRUE
  pdf_document:
    toc: TRUE
    highlight: tango
    number_sections: TRUE
    latex_engine: lualatex
    keep_tex: FALSE
    fig_caption: yes
    fig_crop: no
    fig_height: 2
    fig_width: 3
    toc_depth: 2
always_allow_html: TRUE
urlcolor: blue
linkcolor: black
citecolor: blue
geometry: margin=1in
header-includes: 
  - \usepackage{titling}
  - \usepackage{fancyhdr}
  - \usepackage{graphicx}
  - \usepackage{float}
params:
  resolution_list: '.'
  assay: '.'
  cell_type_label_go: '.'
  condition_value1: '.'
  condition_value2: '.'
  condition_value3: '.'
  root_dir: './'
  PROJECT_NAME: '.'
  PI_NAME: '.'
  TASK_ID: '.'
  PROJECT_LEAD_NAME: '.'
  DEPARTMENT: '.'
  LEAD_ANALYSTS: '.'
  GROUP_LEAD: '.'
  CONTACT_EMAIL: '.'
  PIPELINE: '.'
  START_DATE: '.'
  COMPLETION_DATE: '.'
---

```{r logo-file, echo=FALSE}
attach(params)
# Insert logo on the top of the html report 
logo_file <- file.path(root_dir, "figures", "img", "DNB-BINF-Core-logo.png")
htmltools::img(src = knitr::image_uri(logo_file), alt = "logo", style = "position:absolute; top:0; left:0; padding:0px; height:120px;")
detach(params)
```

\addtolength{\headheight}{2.0cm} 
\fancypagestyle{plain}{} 
\thispagestyle{fancy}
\fancyhead[L]{\includegraphics[height=120px]{`r logo_file`}}
\renewcommand{\headrulewidth}{0pt}

<style type="text/css">
:root {--DNB_BINF_Core_color: #00427B;}

h1.title {margin-top: 130px;
          margin-bottom: 25px;
          font-size: 36px;}

.nobullet li {list-style-type: none;}

.reporthead {font-size: 20px;}

body { /* Normal */
  font-size: 16px;
  font-style: Arial, Helvetica, sans-serif;}

h1 {color: var(--DNB_BINF_Core_color);
    font-size: 28px;
    margin-top: 50px;}

h2 {color: var(--DNB_BINF_Core_color);
    font-size: 20px;}

.list-group-item.active, .list-group-item.active:focus, .list-group-item.active:hover {
  background-color: var(--DNB_BINF_Core_color);}
</style>

<a href="https://wiki.stjude.org/display/CAB">

</a>

\pagebreak

<div class="reporthead"><br/>
**PI: `r params$PI_NAME`**  
**Project: `r params$PROJECT_NAME`**  
Task: `r params$TASK_ID`  
Project Lead(s): `r params$PROJECT_LEAD_NAME`  
Department: `r params$DEPARTMENT`  

<br />  

DNB Bioinformatics Core Analysis Team: 
<br />  

>**Lead Analyst(s): `r params$LEAD_ANALYSTS`**  
>Group Lead: `r params$GROUP_LEAD`  
<br />
>**Contact E-mail:** `r params$CONTACT_EMAIL`  
>**DNB Bioinformatics Core Pipeline:** `r params$PIPELINE`  

Date started: `r params$START_DATE`  
Date completed:  `r params$COMPLETION_DATE`  
Report generated: `r format(Sys.time(), '%H:%M:%S %Z %m/%d/%Y')` \

Reviewed by: _____________________   Date: ____________ \
</div>
\pagebreak
  
# Information about this notebook

Seurat can help find markers that define clusters via differential expression (DE) per cell type. By default, it identifies positive and negative markers of a single cluster (specified in ident.1), compared to all other cells. [FindAllMarkers()](https://satijalab.org/seurat/reference/findallmarkers) automates this process for all clusters, but you can also test groups of clusters vs. each other, or against all cells. 

# Set up

```{r load-library, echo=TRUE}
attach(params)
suppressPackageStartupMessages({
  library(tidyverse)
  library(knitr)
  library(DT)
  library(glue)
  library(Seurat)
}) 
```

# Directories and paths to file Inputs/Outputs

```{r set-dir-and-file-names, echo=TRUE}
analysis_dir <- file.path(root_dir, "analyses", "de-go-analysis") 
module_results_dir <- file.path(analysis_dir, "results")
module_plots_dir <- file.path(analysis_dir, "plots")
data_dir <- file.path(root_dir, "analyses", "cell-types-annotation", "results", "cell_types_annotations_all")

# Input files
data_file <- file.path(data_dir, "seurat_obj_cell_types_annotations_all.rds")

# Create results_dir
results_dir <- file.path(module_results_dir, "01-de-analysis")
if (!dir.exists(results_dir)) {
  dir.create(results_dir)}
```

```{r echo=FALSE, warning=FALSE}
opts_chunk$set(fig.align='center',
               external=TRUE,
               echo=FALSE,
               warning=FALSE,
               fig.pos='H')
a4width <- 8.3
a4height <- 11.7
```

# Read seurat object

First, we will use the seurat object as generated from the pipeline in the `cell-types-annotation` module. 

```{r read-object, echo=TRUE}
seurat_obj <- readRDS(data_file)
DefaultAssay(seurat_obj) <- assay
```

# Differential Expression (DE) Per Cell Type

We save all the tables with Differential Expression (DE) per `r cell_type_label_go` at: <font size="3"><span style="color:blue">./results/01-de-analysis/<FILENAME>.tsv.</span></font>\

## Overall

```{r find-markers-cell-type, echo=TRUE}
Idents(seurat_obj) <- seurat_obj@meta.data[[cell_type_label_go]] # Set Idents(seurat_obj) to match the group.by parameter you will use
table(Idents(seurat_obj)) # View cell identities, get summary table
all.markers <- FindAllMarkers(seurat_obj, assay = assay, logfc.threshold = 0.25, test.use = "wilcox", min.pct = 0.1, return.thresh = 0.01)
write.table(all.markers, file = paste(results_dir, "/", "Res_", resolution_list, "_Markers_all.tsv", sep = ""), sep = "\t", quote = FALSE, row.names = FALSE)
```

```{r find-markers-cell-type-interactive-table, echo=TRUE}
datatable(all.markers, 
          options = list(pageLength = 5, 
                         autoWidth = TRUE, 
                         server = TRUE), 
          filter = "top")
```

## Per `r condition_value1`

```{r find-markers-cell-type-condition1, echo=TRUE}
Idents(seurat_obj) <- interaction(
  seurat_obj@meta.data[[cell_type_label_go]],
  seurat_obj@meta.data[[condition_value1]],
  drop = TRUE)
table(Idents(seurat_obj)) # View cell identities, get summary table
all.markers_condition_value1 <- FindAllMarkers(seurat_obj, assay = assay, logfc.threshold = 0.25, test.use = "wilcox", min.pct = 0.1, return.thresh = 0.01)
write.table(all.markers_condition_value1, file = paste(results_dir, "/", "Res_", resolution_list, "_Markers_all_condition_value1.tsv", sep = ""), sep = "\t", quote = FALSE, row.names = FALSE)
```

```{r find-markers-cell-type-interactive-table-type-condition1, echo=TRUE}
datatable(all.markers_condition_value1, 
          options = list(pageLength = 5, 
                         autoWidth = TRUE, 
                         server = TRUE), 
          filter = "top")
```

## Per `r condition_value2`

```{r find-markers-cell-type-condition2, echo=TRUE}
if (!is.null(condition_value2)) {
  Idents(seurat_obj) <- interaction(
  seurat_obj@meta.data[[cell_type_label_go]],
  seurat_obj@meta.data[[condition_value2]],
  drop = TRUE)
  table(Idents(seurat_obj)) # View cell identities, get summary table
  all.markers_condition_value2 <- FindAllMarkers(seurat_obj, assay = assay, logfc.threshold = 0.25, test.use = "wilcox", min.pct = 0.1, return.thresh = 0.01)
  write.table(all.markers_condition_value2, file = paste(results_dir, "/", "Res_", resolution_list, "_Markers_all_condition_value2.tsv", sep = ""), sep = "\t", quote = FALSE, row.names = FALSE)
}
```

```{r find-markers-cell-type-interactive-table-type-condition2, echo=TRUE}
if (!is.null(condition_value2)) {
  datatable(all.markers_condition_value2, 
            options = list(pageLength = 5, 
                           autoWidth = TRUE, 
                           server = TRUE), 
            filter = "top")
}
```

## Per `r condition_value3`

```{r find-markers-cell-type-condition3, echo=TRUE}
if (!is.null(condition_value2)) {
  Idents(seurat_obj) <- interaction(
  seurat_obj@meta.data[[cell_type_label_go]],
  seurat_obj@meta.data[[condition_value3]],
  drop = TRUE)
  table(Idents(seurat_obj)) # View cell identities, get summary table
  all.markers_condition_value3 <- FindAllMarkers(seurat_obj, assay = assay, logfc.threshold = 0.25, test.use = "wilcox", min.pct = 0.1, return.thresh = 0.01)
  write.table(all.markers_condition_value3, file = paste(results_dir, "/", "Res_", resolution_list, "_Markers_all_condition_value3.tsv", sep = ""), sep = "\t", quote = FALSE, row.names = FALSE)
}
```

```{r find-markers-cell-type-interactive-table-type-condition3, echo=TRUE}
if (!is.null(condition_value3)) {
  datatable(all.markers_condition_value3, 
            options = list(pageLength = 5, 
                           autoWidth = TRUE, 
                           server = TRUE), 
            filter = "top")
}
```

```{r echo=FALSE}
detach(params)
```

\pagebreak

# Session Info

```{r echo=FALSE}
sessionInfo()
```
